File: 12.benchmark/inversion.rb
def get_matrix_inverse(matrix, n)
    augmented = Array.new(n) { Array.new(2 * n, 0.0) }
  
    # Create the augmented matrix [matrix | identity]
    n.times do |i|
      n.times do |j|
        augmented[i][j] = matrix[i][j]
        augmented[i][j + n] = (i == j) ? 1.0 : 0.0
      end
    end
  
    # Perform Gauss-Jordan elimination
    n.times do |i|
      # Make the diagonal contain all 1's
      diag_element = augmented[i][i]
      (2 * n).times do |j|
        augmented[i][j] /= diag_element
      end
  
      # Make the other elements in the current column 0
      n.times do |k|
        next if k == i
        factor = augmented[k][i]
        (2 * n).times do |j|
          augmented[k][j] -= factor * augmented[i][j]
        end
      end
    end
  
    # Extract the inverse matrix from the augmented matrix
    inverse = Array.new(n) { Array.new(n, 0.0) }
    n.times do |i|
      n.times do |j|
        inverse[i][j] = augmented[i][j + n]
      end
    end
  
    inverse
  end
  
  def main
    # Read the size of the matrix
    puts "Enter the size of the matrix: "
    n = gets.to_i
  
    matrix = Array.new(n) { Array.new(n, 0.0) }
  
    # Input each row of the matrix
    puts "Enter the elements of the matrix row by row:"
    n.times do |i|
      row = gets.split.map(&:to_f)
      n.times do |j|
        matrix[i][j] = row[j]
      end
    end
  
    inverse = get_matrix_inverse(matrix, n)
  
    # Print the inverse matrix
    puts "Inverse matrix:"
    inverse.each do |row|
      puts row.map { |val| format('%.6f', val) }.join(' ')
    end
  end
  
  main

File: 12.benchmark/inversion.rs
use std::io::{self};

fn get_matrix_inverse(matrix: &mut Vec<Vec<f64>>, n: usize) -> Vec<Vec<f64>> {
    let mut augmented: Vec<Vec<f64>> = vec![vec![0.0; 2 * n]; n];

    // Create the augmented matrix [matrix | identity]
    for i in 0..n {
        for j in 0..n {
            augmented[i][j] = matrix[i][j];
            augmented[i][j + n] = if i == j { 1.0 } else { 0.0 };
        }
    }

    // Perform Gauss-Jordan elimination
    for i in 0..n {
        // Make the diagonal contain all 1's
        let diag_element = augmented[i][i];
        for j in 0..2 * n {
            augmented[i][j] /= diag_element;
        }

        // Make the other elements in the current column 0
        for k in 0..n {
            if k != i {
                let factor = augmented[k][i];
                for j in 0..2 * n {
                    augmented[k][j] -= factor * augmented[i][j];
                }
            }
        }
    }

    // Extract the inverse matrix from the augmented matrix
    let mut inverse: Vec<Vec<f64>> = vec![vec![0.0; n]; n];
    for i in 0..n {
        for j in 0..n {
            inverse[i][j] = augmented[i][j + n];
        }
    }

    inverse
}

fn main() {
    let stdin = io::stdin();
    let mut input = String::new();

    // println!("Enter the size of the matrix (n x n): ");
    stdin.read_line(&mut input).expect("Failed to read line");
    let n: usize = input.trim().parse().expect("Please enter a number");

    let mut matrix: Vec<Vec<f64>> = vec![vec![0.0; n]; n];

    // Input each row of the matrix
    println!("Enter the elements of the matrix row by row:");
    for i in 0..n {
        input.clear();
        stdin.read_line(&mut input).expect("Failed to read line");
        let row: Vec<f64> = input
            .trim()
            .split_whitespace()
            .map(|x| x.parse().expect("Please enter a number"))
            .collect();
        for j in 0..n {
            matrix[i][j] = row[j];
        }
    }

    let inverse = get_matrix_inverse(&mut matrix, n);

    // Print the inverse matrix
    println!("Inverse matrix:");
    for i in 0..n {
        for j in 0..n {
            print!("{:.6} ", inverse[i][j]);
        }
        println!();
    }
}